### PR TITLE
3D GANs 

### PR DESCRIPTION
**Pull request by Psych17101**
Implementation of 3 different 3D GANs applied to 3D Heat Condution: cGAN 3D, SliceGAN, and MV-VAE-GAN.

### CODE REVIEW

File: engiopt/cgan_cnn_3d/sweep_cgan_cnn_3d_grid.yaml [hunk 1]
```diff
@@ -0,0 +1,25 @@
+method: grid
+metric:
+  name: dpp
+  goal: minimize
```
Comment (ffelten) [inline engiopt/cgan_cnn_3d/sweep_cgan_cnn_3d_grid.yaml:4]: This should be maximized no?

File: engiopt/cgan_cnn_3d/cgan_cnn_3d.py [hunk 2]
```diff
@@ -0,0 +1,671 @@
...
+
+        # Final classification
+        return self.final_conv(h)  # -> (B, out_channels, 1, 1, 1)
+
+
+def compute_gradient_penalty(discriminator, real_samples, fake_samples, conds, device, lambda_gp=20.0):
+    """Calculates the gradient penalty loss for WGAN GP"""
+    batch_size = real_samples.size(0)
+    # Random weight term for interpolation between real and fake samples
+    alpha = th.rand(batch_size, 1, 1, 1, 1, device=device)
```
[the remark this answers is not in the document]
Comment (Psych17101) [reply engiopt/cgan_cnn_3d/cgan_cnn_3d.py:324]: in my paper "improving training with Wasserstein GANS", the gradient penalty is computed by interpolating with random alpha sampled uniformily. Makes sure that the penalty it on a straight line and not just the mid point.

File: engiopt/slice_gan_3d/slice_gan_3d.py [hunk 3]
```diff
@@ -0,0 +1,906 @@
...
+                optimizer_generator.zero_grad()
+
+                # Sample latent vector (Line 17)
+                z = th.randn((batch_size, args.latent_dim), device=device)
+                # Generate 3D volume (Line 18)
+                fake_volumes_tensor = slice_generator(z, conds_expanded)
+                fake_volumes = {"generated": fake_volumes_tensor}  # Or appropriate dict key
+
+                total_g_loss = th.tensor(0.0, device=device, requires_grad=True)
+                # For each axis a = 1, 2, 3 (Line 19)
```
[the remark this answers is not in the document]
Comment (Psych17101) [reply engiopt/slice_gan_3d/slice_gan_3d.py:710]: Used for my reference/check based on lines from the algorithm in the publication, should they be removed?

**Review by ffelten — APPROVED** [review]
